source/core/slang-list.h

```diff
@@ -43,126 +43,136 @@ namespace Slang
         }
     };
 
-	template<typename T, typename TAllocator>
-	class AllocateMethod
-	{
-	public:
-		static inline T* allocateArray(Index count)
-		{
-			TAllocator allocator;
-			T * rs = (T*)allocator.allocate(count * sizeof(T));
-			Initializer<T, std::is_pod<T>::value>::initialize(rs, count);
-			return rs;
-		}
-		static inline void deallocateArray(T* ptr, Index count)
-		{
-			TAllocator allocator;
-			if (!std::is_trivially_destructible<T>::value)
-			{
-				for (Index i = 0; i < count; i++)
-					ptr[i].~T();
-			}
-			allocator.deallocate(ptr);
-		}
-	};
-
-	template<typename T>
-	class AllocateMethod<T, StandardAllocator>
-	{
-	public:
-		static inline T* allocateArray(Index count)
-		{
-			return new T[count];
-		}
-		static inline void deallocateArray(T* ptr, Index /*bufferSize*/)
-		{
-			delete [] ptr;
-		}
-	};
-
-
-	template<typename T, typename TAllocator = StandardAllocator>
-	class List
-	{
+    template<typename T, typename TAllocator>
+    class AllocateMethod
+    {
+    public:
+        static inline T* allocateArray(Index count)
+        {
+            TAllocator allocator;
+            T * rs = (T*)allocator.allocate(count * sizeof(T));
+            Initializer<T, std::is_pod<T>::value>::initialize(rs, count);
+            return rs;
+        }
+        static inline void deallocateArray(T* ptr, Index count)
+        {
+            TAllocator allocator;
+            if (!std::is_trivially_destructible<T>::value)
+            {
+                for (Index i = 0; i < count; i++)
+                    ptr[i].~T();
+            }
+            allocator.deallocate(ptr);
+        }
+    };
+
+    template<typename T>
+    class AllocateMethod<T, StandardAllocator>
+    {
+    public:
+        static inline T* allocateArray(Index count)
+        {
+            return new T[count];
+        }
+        static inline void deallocateArray(T* ptr, Index /*bufferSize*/)
+        {
+            delete [] ptr;
+        }
+    };
+
+    // List is container of values of a type held consecutively in memory (much like std::vector)
+    // 
+    // Note that in this implementation, the underlying memory is backed via an allocation of T[capacity]
+    // This means that all values have to be in a valid state *even if they are not used* (ie indices >= m_count must be valid)
+    //
```

Comment:
Should we try to fix this behavior? It seems to be obviously wrong as soon as `List<...>` gets used together with types that have destructors, which is something we do all the time right now. There is merit in a similar type that is designed to only work with POD types, but that is also one of those things that "modern C++" at least has an answer for.

Reply:
Yes I think we should fix this aspect, and make it not backed via T[], as that just wastes memory and assumes everything is initialized, even when it doesn't need to be.